### PR TITLE
Added unit tests, dockerfile, minor code changes. 

### PR DESCRIPTION
**Pull request by HumairAK**
Hello! The following changes are included in this PR:

**Tests**
The testing framework used is pytest. You can find all the dependencies used in the requirements-tests.txt and the tox.ini (under deps). Tests can be ran by simply invoking the `tox` command in the terminal (in the directory in which tox.ini resides) after cloning. After the first tox invocation, any consecutive test runs will run faster. Some resource files for testing purposes are also included. 

**Docker-CI file**
Docker file was submitted by [dsoukhov](https://github.com/dsoukhov) for CI purposes. 

**Changes to code base** 
The following changes were made: 
- Minor code changes to follow pep8 specification. 
- Refactored code in `workloop` method in `worker.py` to make it more testable (simply moved to a new method `update_model`). 
- `options() `global is now defined within a global configs file called `conf.py `and initialized in `app.py.` Initially `options()` was imported in `defaultcontrollers.py` by the calling script via the 'from __main__ import options' command. This setup made it impossible to unit test `defaultcontrollers.py.` 

Please do let me know if there are any issues at your leisure, 
Thanks!

### CODE REVIEW

**Comment by zak-hassan** [comment]
I spoke with @elmiko and asked him to review this patch. @HumairAK thanks for your contribution.

**Review by elmiko — CHANGES_REQUESTED** [review]
this looks really nice @HumairAK, i have a couple questions and one request.

i noticed that building the container image from the Dockerfile-ci looks like it is running the tests during the build phase, is this by design?

i would have expected the Dockerfile-ci to produce an image that runs the tests, is there something i am missing about our CI setup?

would you mind adding text to the README.md file about how to run the tests

File: Dockerfile-ci [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+RUN pip install -r requirements.txt \
+ && pip wheel -r wheel-requirements.txt -w . \
+ && mv pymongo*.whl pymongo.zip \
+ && pip install -r requirements-tests.txt \
+ && pip install -e  $SPARK_HOME/python/
+
+ENV PYTHONPATH=$SPARK_HOME/python:$PYTHONPATH
+
+RUN export PYTHONPATH=$PYTHONPATH:$(ls $SPARK_HOME/python/lib/py4j*.zip | head -1) \
+ && tox
```
Comment (elmiko) [inline Dockerfile-ci:15]: this is running the test during image creation, is that the expected behavior?

**Comment by dsoukhov** [comment]
I designed the dockerfile-ci to run the test during image creation. I have a jenkins build set up on et24 that uses it for automated testing. If a test fails, the build also fails. 

I'll add some info to the readme on how to run the tests.

**Comment by elmiko** [comment]
excellent, thanks @dsoukhov !

**Comment by elmiko** [comment]
thanks for the updates @HumairAK and @dsoukhov lgtm!

**Comment by willb** [comment]
Thanks, guys!

**Pull request merged**